private/bufpkg/bufmodule/bufmodulestore/module_data_store.go

```diff
@@ -167,14 +183,33 @@ func (p *moduleDataStore) getModuleDataForModuleKey(
 			return nil, err
 		}
 	} else {
-		moduleCacheBucket, err = p.getReadWriteBucketForDir(moduleKey)
-		// Not checking for fs.ErrNotExist. Function only returns error on actual system error.
+		dirPath, err := getModuleDataStoreDirPath(moduleKey)
 		if err != nil {
 			return nil, err
 		}
+		p.logDebugModuleKey(
+			moduleKey,
+			"module data store dir read write bucket",
+			zap.String("dirPath", dirPath),
+		)
+		moduleCacheBucket = storage.MapReadWriteBucket(p.bucket, storage.MapOnPrefix(dirPath))
+		// Only attempt to get a file lock when storing individual files
+		if p.filelocker != nil {
+			unlocker, err := p.filelocker.RLock(ctx, dirPath+"/"+externalModuleDataLockFileName)
+			if err != nil {
+				p.logger.Debug("failed_get_rlock", zap.Error(err))
+				return nil, err
+			}
+			defer func() {
+				if err := unlocker.Unlock(); err != nil {
+					p.logger.Debug("failed_get_runlock", zap.Error(err))
+					retErr = multierr.Append(retErr, err)
+				}
+			}()
+		}
 	}
 	defer func() {
-		if retErr != nil {
+		if retErr != nil && !errors.Is(err, fs.ErrNotExist) {
```

Comment:
I don't know if `fs.ErrNotExist` can be returned by any other calls other than the `storage.ReadPath` below - if so we might want to tweak this.

Reply:
Yes, those calls can return `fs.ErrNotExist`, but we only attempt to read those paths if they are parsed from a valid `module.yaml`. Might still make sense to call `deleteInvalidModuleData` in those cases, so let me rework this.